pkg/chunk/elastic/es_index_client.go

```diff
@@ -0,0 +1,344 @@
+	return &elasticReadBatchIterator{
+		readBatch: r,
+	}
+}
+
+type elasticReadBatchIterator struct {
+	consumed bool
+	readBatch
+}
+
+func (r *elasticReadBatchIterator) Next() bool {
+	if r.consumed {
+		return false
+	}
+	r.consumed = true
+	return true
+}
+
+func (r *elasticReadBatchIterator) RangeValue() []byte {
+	return r.rangeValue
+}
+
+func (r *elasticReadBatchIterator) Value() []byte {
+	return r.value
+}
+
+func (e *esClient) QueryPages(ctx context.Context, queries []chunk.IndexQuery, callback func(chunk.IndexQuery, chunk.ReadBatch) (shouldContinue bool)) error {
+	return chunk_util.DoParallelQueries(ctx, e.query, queries, callback)
+}
+
+func (e *esClient) query(ctx context.Context, query chunk.IndexQuery, callback func(chunk.ReadBatch) (shouldContinue bool)) error {
+	var rangeQuery *elastic.RangeQuery
+	var valueTermQuery *elastic.TermQuery
+
+	level.Debug(util.Logger).Log("msg", fmt.Sprintf(
+		"hash [%s], rangeValuePrefix [%s], rangeValueStart [%s]", query.HashValue, query.RangeValuePrefix, query.RangeValueStart))
+	hashTermQuery := elastic.NewTermQuery("hash", query.HashValue)
+	switch {
+	case len(query.RangeValuePrefix) > 0 && query.ValueEqual == nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValuePrefix)).
+			Lt(string(query.RangeValuePrefix) + null)
+
+	case len(query.RangeValuePrefix) > 0 && query.ValueEqual != nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValuePrefix)).
+			Lt(string(query.RangeValuePrefix) + null)
+		valueTermQuery = elastic.NewTermQuery("value", query.ValueEqual)
+
+	case len(query.RangeValueStart) > 0 && query.ValueEqual == nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValueStart))
+
+	case len(query.RangeValueStart) > 0 && query.ValueEqual != nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValueStart))
+		valueTermQuery = elastic.NewTermQuery("value", query.ValueEqual)
+
+	case query.ValueEqual != nil:
+		valueTermQuery = elastic.NewTermQuery("value", query.ValueEqual)
+
+	case query.ValueEqual == nil:
+		break
+	}
+
+	exists, err := e.client.IndexExists(query.TableName).Do(ctx)
+	if err != nil {
+		// Handle error
+		fmt.Println(exists)
+	}
+
+	// Search with a term query
+	baseQuery := e.client.Search().
+		Index(query.TableName).
+		Query(hashTermQuery)
+	if valueTermQuery != nil {
+		baseQuery = baseQuery.Query(valueTermQuery)
+	}
+	if rangeQuery != nil {
+		baseQuery = baseQuery.Query(rangeQuery)
+	}
+
+	// Search with a term query
+	searchResult, err := baseQuery.
+		Sort("range", true). // sort by "range" field, ascending
+		From(0).Size(maxFetchDocs).
+		Pretty(true). // pretty print request and response JSON
+		Do(ctx)       // execute
+
+	if searchResult == nil || searchResult.Hits == nil {
+		return nil
+	}
+
+	if err != nil {
+		// Handle error
+		level.Error(util.Logger).Log("msg", fmt.Sprintf("Query in index %s met error!", query.TableName))
+		panic(err)
```

Comment:
Panicing seems to be a bad behaviour during a query.

Reply:
Indeed. I forgot this.